allennlp/semparse/worlds/text2sql_world.py

```diff
@@ -38,6 +65,31 @@ def get_action_sequence_and_all_actions(self, query: List[str]) -> Tuple[List[st
         action_sequence = sql_visitor.parse(" ".join(query)) if query else []
         return action_sequence, sorted_actions
 
-    @staticmethod
-    def is_global_rule(nonterminal: str) -> bool: # pylint: disable=unused-argument
+    def _initialize_grammar_dictionary(self, grammar_dictionary: Dict[str, List[str]]) -> Dict[str, List[str]]:
+        # Add all the table and column names to the grammar.
+        if self.schema:
+            update_grammar_with_tables(grammar_dictionary, self.schema)
+
+            if self.cursor is not None and not self.use_prelinked_entities:
+                # Now if we have strings in the table, we need to be able to
+                # produce them, so we find all of the strings in the tables here
+                # and create production rules from them. We only do this if
+                # we haven't pre-linked entities, because if we have, we don't
+                # need to be able to generate the values - just the placeholder
+                # symbols which link to them.
+                grammar_dictionary["number"] = []
+                grammar_dictionary["string"] = []
+
+                update_grammar_with_table_values(grammar_dictionary, self.schema, self.cursor)
+            else:
+                # TODO(Mark): The grammar can be tightened here if we don't need to
+                # produce concrete values.
+                pass
+
+        return grammar_dictionary
+
+    def is_global_rule(self, production_rule: str) -> bool:
+        if self.use_prelinked_entities:
+            if "value" in production_rule and production_rule[-4].isnumeric():
```

Comment:
I think it would be helpful to add a comment here: we are checking `-4` so that it is not a global rule if we see the `0` in the a rule like ``'value -> ["\'city_name0\'"]'``, right?